test/python/dagcircuit/test_dagcircuit.py

```diff
@@ -248,6 +248,140 @@ def test_addding_individual_bit(self):
+        clbit = Clbit()
+        self.dag.add_qreg(qreg)
+        self.dag.add_creg(creg0)
+        self.dag.add_creg(creg1)
+        self.dag.add_creg(creg2)
+        self.dag.add_clbits([clbit])
+
+        self.qreg = qreg
+        self.creg0 = creg0
+        self.creg1 = creg1
+        self.creg1_alias = creg2
+        self.individual_clbit = clbit
+
+        self.original_cregs = [self.creg0, self.creg1, self.creg1_alias]
+
+        # skip adding bits of creg1_alias since it's just an alias for creg1
+        self.original_clbits = [
+            b for reg in self.original_cregs if reg != self.creg1_alias for b in reg
+        ] + [self.individual_clbit]
+
+    def assert_cregs_is(self, cregs, excluding=None):
+        """Assert test DAG cregs match the expected values.
+
+        Args:
+            cregs (Iterable(ClassicalRegister)): the classical registers to expect
+            excluding (Set(ClassicalRegister)): classical registers to remove from
+            ``cregs`` before the comparison."""
+        if excluding is None:
+            excluding = {}
+        self.assertDictEqual(
+            self.dag.cregs, {creg.name: creg for creg in cregs if creg not in excluding}
+        )
+
+    def assert_clbits_is(self, clbits, excluding=None):
+        """Assert test DAG clbits match the expected values.
+
+        Args:
+            clbits (Iterable(Clbit)): the classical bits to expect
+            excluding (Set(ClassicalRegister)): classical bits to remove from
+            ``clbits`` before the comparison."""
+        if excluding is None:
+            excluding = {}
+        self.assertListEqual(self.dag.clbits, list(b for b in clbits if b not in excluding))
+
+    def test_remove_idle_creg(self):
+        """Removing an idle classical register removes just the register."""
+        self.dag.remove_cregs(self.creg0)
+
+        self.assert_cregs_is(self.original_cregs, excluding={self.creg0})
+        self.assert_clbits_is(self.original_clbits)
+
+    def test_remove_busy_creg(self):
+        """Classical registers with both busy and idle underlying bits
+        can be removed, keeping underlying bits."""
+        self.dag.apply_operation_back(Measure(), [self.qreg[0]], [self.creg0[0]])
+        self.dag.remove_cregs(self.creg0)
+
+        # creg removal always succeeds
+        self.assert_cregs_is(self.original_cregs, excluding={self.creg0})
+
+        # creg0 is gone, leaving both creg0[0] and creg0[1] unreferenced.
+        # however, only creg0[1] is idle, and thus removed.
+        self.assert_clbits_is(self.original_clbits)
+
+    def test_remove_cregs_shared_bits(self):
+        """Removing a classical register does not remove other classical
+        registers that are simply aliases of it."""
+        self.dag.remove_cregs(self.creg1)
+
+        # Only creg1 should be deleted, not its alias
+        self.assert_cregs_is(self.original_cregs, excluding={self.creg1})
+        self.assert_clbits_is(self.original_clbits)
```

Comment:
I'm not going to ask for changes to the whole test file here, but using `setUp` to do all the work makes the individual test cases a bit hard to read - I can't verify from just looking at this test what's meant by "`self.creg1` is an alias", so it's not entirely clear to me what's being tested.  

Obviously if I look at the `setUp` I can work it out, but it's hard to see at a glance - maybe I'm in the minority, but to me, set-up repetition in a test suite isn't really a bad thing, because it makes it clear _exactly_ what's being tested, and the local test site (and ensures that set-up for other unrelated tests doesn't influence results).

Reply:
I will keep this in mind for next time. I'm not sure how I feel about it either—I simply copied this pattern from other test classes. I suppose if there's a test failure, it wouldn't be too hard to take a look at both the specific test as well as the set up method. If everything starts failing, maybe this is a much bigger pain 😄.